enhancements/ingress/global-options-enable-hsts.md

```diff
@@ -0,0 +1,457 @@
+```go
+	const RouterHSTSHeaderList        = "ROUTER_HSTS_LIST"
+
+       if scope ==  LimitedHSTSHeaderPolicyScope {
+            env = append(env, corev1.EnvVar{Name: RouterHSTSHeaderScope, Value: LimitedHSTSHeaderPolicyScope})
+            env = append(env, corev1.EnvVar{Name: RouterHSTSHeaderList, Value: ci.Spec.HTTPHeaders.HSTSHeaderPolicy.InScopeDomains})
+       }
+```
+As shown in this example, the HSTS Policy scope is for a limited list of two domains, expires after one year, includes
+subdomains, and requests preload:
+```yaml
+apiVersion: operator.openshift.io/v1
+kind: IngressController
+metadata:
+  name: default
+  namespace: openshift-ingress-operator
+spec:
+  httpHeaders:
+    hstsHeaderPolicy:
+      scope: Limited
+      inScopeDomains:
+        - domainA
+        - domainB
+      maxAgeSeconds: 31536000
+      hstsDirectives:
+        - includeSubDomains
+        - preload
+```
+
+With the addition of processing for `ROUTER_HSTS_SCOPE` of `limited`, and `ROUTER_HSTS_LIST`, the HAProxy template code
+snippet above will gain a few new variables:
+```gotemplate
+{{- $router_hsts := env "ROUTER_DEFAULT_HSTS" "" }}
+{{- $router_hsts_scope := env "ROUTER_HSTS_SCOPE" "" }}
+{{- $router_hsts_list := env "ROUTER_HSTS_LIST" }}
+...
+{{- if matchValues (print $cfg.TLSTermination) "edge" "reencrypt" }}
+  {{- /* Get the HSTS header value using the default and the annotations in the cfg */ }}
+    {{- with $hsts := hstsHeaderValue $router_global_hsts $router_hsts_scope $cfg $router_hsts_list }}
+      http-response set-header Strict-Transport-Security {{$hsts}}
+    {{- end }}{{/* hstsHeaderValue */}}
+{{- end }}{{/* is "edge" or "reencrypt" */}}
+```
+
+#### Disable HSTS per-route when it is enabled globally
+As mentioned previously, the `Strict-Transport-Security` annotation called
+`haproxy.router.openshift.io/hsts_header` has a required directive `max-age` that determines
+its lifespan.  A per-route override on globally enabled HSTS can be indicated by setting the `max-age`
+value in the route's annotation to zero, as shown in this example:
+```yaml
+metadata:
+  annotations:
+    haproxy.router.openshift.io/hsts_header: max-age=0
+```
+
+The HAProxy template remains the same, and the promised new function `hstsHeaderValue` will be added:
+
+```go
+// If the override annotation string is set, return it - the annotation overrides the default
+// Otherwise, if the scope is limited and the domain is in the list - return the default
+// Finally, if the scope is all - return the default
+// Otherwise, return the empty string
+func hstsHeaderValue(default_header string, hsts_scope string, config ServiceAliasConfig, hsts_list []string) string {
+    wantLimited := false
+
+	if override, ok := config.Annotations["haproxy.router.openshift.io/hsts_header"]; ok {
+        return override
+    } else if hsts_scope == LimitedHSTSHeaderPolicyScope {
+        for domain := range hsts_list{
+            if strings.Contains(config.Host, domain) {
```

Comment:
It would be better to parse `ROUTER_HSTS_LIST` into a `sets.String` value in the template plugin setup when the router starts in order to avoid doing a lot of string manipulation or looping when generating `haproxy.config`, but we can work out those implementation details later on.  

Using the host from the `ServiceAliasConfig` is a terrific idea—much simpler than writing a map and using HAProxy acls.

Reply:
I'll keep it a semi-colon delimited string to be passed in as an env variable, and do the parsing in hstsHeaderValue.
Thanks for all your comments!